### PR TITLE
feat: warn user to hold regen for tx fees

### PR DESCRIPTION
**Pull request by flagrede**
## Description

Closes: regen-network/regen-registry#1270

- check if the user has enough regen for the tx fees
- display a modal if the user has not enough regen
- create an error system with a single TxModalError instance
  - static errors info defined in [config/errors](https://github.com/regen-network/regen-web/pull/1660/files#diff-b92c3b4f81b7743c9d28cb7b759f04aa8bd1890ce27f172a309b1fcb70c3a92d)
  - dynamic info  defined in [globalContext](https://github.com/regen-network/regen-web/pull/1660/files#diff-1456997fb750053d18d828a6fc23cea0af74bd9454e3c3475aaa451e4614f05b)
  - dynamic overrides static if both keys exist.

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] provided a link to the relevant issue or specification
- [x] provided instructions on how to test
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] once the PR is closed, set up backport PRs for `redwood`, `hambach` and optionally `master` (see below)

#### Setting up backport PRs

After merging your PR to `dev`, set up backports by doing the following:

1. If your branch does not have merge commits, add the following comment to
   your PR:

- If PR is a new feature: `@Mergifyio backport redwood hambach`.
- If PR is a bug fix: `@Mergifyio backport redwood hambach master`.

2. If your branch does have merge commits:

   a. Pull latest `dev`, `hambach` and `redwood` branches

   b. Create new branches for backports and merge `dev` (replace `<PR#>` with your PR #)

   ```
   git checkout -b hambach-backport-<PR#> hambach
   git merge dev
   git push origin hambach-backport-<PR#>
   git checkout -b redwood-backport-<PR#> redwood
   git merge dev
   git push origin redwood-backport-<PR#>`
   ```

   c. Open new PRs in regen-web targeting `hambach` and `redwood`, respectively.

   d. If PR is a bug fix repeat the same process for `master` branch.

### How to test

1. https://deploy-preview-1660--regen-registry.netlify.app/
2. connect with an account with no regen and try to do a tx on one of these actions:
  - Buy
  - Create sell order
  - Cancel sell order
  - Send
  - Retire
  - Put in basket
  - Take from basket

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: web-components/src/components/modal/TxErrorModal.tsx [hunk 1]
```diff
@@ -15,22 +15,22 @@ const TxErrorModal: React.FC<React.PropsWithChildren<Props>> = ({
...
+    ? ([
+        {
+          label: 'error',
+          value: {
+            name: error,
+          },
+          color: 'error.main',
+        },
+      ] as Item[])
+    : undefined;
```
Comment (flagrede) [inline web-components/src/components/modal/TxErrorModal.tsx:28]: Prevent sending a cardItems when there's no error.

File: web-components/src/components/modal/TxModal.tsx [hunk 2]
```diff
@@ -132,38 +135,44 @@ const TxModal: React.FC<React.PropsWithChildren<TxModalProps>> = ({
...
-        ))}
-        {txHash && (
-          <CardItem
-            label="hash"
-            value={{ name: truncate(txHash), url: txHashUrl }}
-            linkComponent={linkComponent}
-          />
-        )}
-      </Card>
+      {hasCardItems && (
```
Comment (flagrede) [inline web-components/src/components/modal/TxModal.tsx:145]: The only change here is hiding the block when there are no card items.

File: web-registry/src/components/organisms/RegistryLayout/RegistryLayout.TxErrorModal.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  return (
+    <>
+      <TxErrorModal
+        error={txError}
+        open={!!errorCode}
+        onClose={resetTxModalError}
+        txHash={txHash ?? ''}
+        txHashUrl={txHashUrl}
+        title={findFirstNonEmptyString([title, error.title])}
```
Comment (flagrede) [inline web-registry/src/components/organisms/RegistryLayout/RegistryLayout.TxErrorModal.tsx:44]: The idea here is to use data from the context first and fall back to the static config.

File: web-registry/src/components/organisms/RegistryLayout/RegistryLayout.TxErrorModal.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      buttonLink,
+      buttonTitle,
+      cardItems,
+      cardTitle,
+      description,
+      title,
+      txError,
+      txHash,
+    },
+  ] = useGlobalStore(store => store['errorModal']);
```
Comment (flagrede) [inline web-registry/src/components/organisms/RegistryLayout/RegistryLayout.TxErrorModal.tsx:29]: Optional values set alongside an error.

**Comment by erikalogie** [comment]
@flagrede this looks great!

File: web-registry/src/components/organisms/WalletButton/WalletButton.tsx [hunk 5]
```diff
@@ -29,6 +33,15 @@ const WalletButton: React.FC = () => {
     useState<WalletModalState>('wallet-select');
   const isConnected = loaded ? !!wallet?.address : null;
 
+  // Populate cache with user balance once connected
+  useQuery(
+    getAllBalancesQuery({
```
Comment (blushi) [inline web-registry/src/components/organisms/WalletButton/WalletButton.tsx:38]: you could just use the `Balance` query which would avoid using `find` afterwards

also, ideally we could also refetch the balance as soon as we try to initiate a new tx
indeed right now, if I top up the regen account balance and then retry, I still get the warning, we need a page refresh to clear the cache and make it disappear

Comment (flagrede) [reply web-registry/src/components/organisms/WalletButton/WalletButton.tsx]: I swapped the query to `Balance` and refetch the cache before checking the balance for the fees.

**Review by victordelval — APPROVED** [review]
a super nice solution, thanks @flagrede! Getting familiar with the context + subscription pattern :muscle:

**Review by blushi — APPROVED** [review]

**Comment by flagrede** [comment]
@Mergifyio backport redwood hambach

**Pull request merged**